### DISCUSSION TITLE
Command Binding of a button inside a DataGridTemplateColumn

### DISCUSSION THREAD

**Discussion by xaberue** [dotnet/wpf#9240] [Q&A] [answered]
Hi all!

I'm struggling trying to bind a command to a button inside of a DataGridTemplateColumn.

Having a look into my problem, I came across some links that I've tried without success:
- https://stackoverflow.com/questions/15494226/cannot-find-source-for-binding-with-reference-relativesource-findancestor
- https://stackoverflow.com/questions/59967164/cannot-find-source-for-binding-with-reference-in-nested-style-datagrid-datatem

This is the code of for the DataGridTemplateColumn

```xaml
<DataGridTemplateColumn Header="View">
    <DataGridTemplateColumn.CellTemplate>
        <DataTemplate>
            <ui:Button Appearance="Secondary" Icon="{ui:SymbolIcon Eye28}"
                Command="{Binding Path=ViewSecretCommand, RelativeSource={RelativeSource AncestorType={x:Type local:SecretsPageViewModel}}}"/>
        </DataTemplate>
    </DataGridTemplateColumn.CellTemplate>
</DataGridTemplateColumn>
```

and this is the command that I'm trying to bind, I'm using the MVVM CommunityToolkit:

```csharp
 [RelayCommand]
 private async Task OnViewSecret(SecretDto secret)
 {
    //TODO
    await Task.CompletedTask;
 }
```

The error that I'm getting is the following:

> Cannot find source: RelativeSource FindAncestor, AncestorType='ChustaSoft.Vaulture.UI.Pages.SecretsPageViewModel', AncestorLevel='1'.

![image](https://github.com/dotnet/wpf/assets/55638917/994964b3-62c8-4d26-81e4-46f5dfa73301)

For sure I'm missing something but not clear what is it.

Please find the entire code [here](https://github.com/ChustaSoft/Vaulture/blob/main/ChustaSoft.Vaulture.UI/Pages/SecretsPage.xaml)

Many thanks in advance!

**Comment by AndrewKeepCoding** [comment, marked as answer]
Try the following methods:

## A
- `x:Name` your `Page` or `Window`, "Root" for example.
- Use the name above to bind the command via the `DataContext`:

```xaml
<Button Command="{Binding ElementName=Root, Path=DataContext.ViewSecretCommand}" .../>
```

## B
- For the `AncestorType`, target a control that is outside of the `DataTemplate`s. I guess in your case, it's the `ItemsControl`.
- Be specific with the [AncestorLevel](https://learn.microsoft.com/en-us/dotnet/api/system.windows.data.relativesource.ancestorlevel?wt.mc_id=MVP_356303).
- `SecretsPageViewModel` is in a `DataContext`. So, your binding should be:

```xaml
<Button Command="{Binding DataContext.ViewSecretCommand, RelativeSource...}" .../>
```

Some links that might help:
- [RelativeSource MarkupExtension](https://learn.microsoft.com/en-us/dotnet/desktop/wpf/advanced/relativesource-markupextension?view=netframeworkdesktop-4.8?wt.mc_id=MVP_356303)
- https://stackoverflow.com/a/15755948/2411960
- https://stackoverflow.com/a/12360603/2411960

**Reply by xaberue** [reply]
Hi @AndrewKeepCoding !!

**Option A** worked perfectly fine, I tried the second as well, but it didn't work for me, maybe I'm missing (and not understanding) something about this. I'm struggling trying to understand the binding context and how to properly define the proper ancestor level.

Something that I would like to know as well is if there's any advantage in using one option over the other. If the option B is significantly better, I can spend an extra time looking into it.

Many thanks a lot, you solved my problem!!

**Reply by miloush** [reply]
`B` is going to traverse the tree at runtime until it finds it the matching element and then resolve the path. `A` is potentially a dictionary lookup by name. I would go for `A` whenever possible for performance reasons.

**Reply by xaberue** [reply]
thanks you a lot @miloush! option A ftw then! Thanks both!
